lib/which_formula.rb

```diff
@@ -2,16 +2,39 @@
 # frozen_string_literal: true
 
 require "formula"
+require "api"
+require "utils/curl"
 
 module Homebrew
   module WhichFormula
-    LIST_PATH = T.let(File.expand_path("#{File.dirname(__FILE__)}/../executables.txt").freeze, String)
+    ENDPOINT = "internal/executables.txt"
+    DATABASE_FILE = T.let((Homebrew::API::HOMEBREW_CACHE_API/ENDPOINT).freeze, Pathname)
 
     module_function
 
+    sig { params(skip_update: T::Boolean).void }
+    def download_and_cache_executables_file!(skip_update: false)
+      if DATABASE_FILE.exist? && !DATABASE_FILE.empty? &&
+         (skip_update || (Time.now - Homebrew::EnvConfig.api_auto_update_secs.to_i) < DATABASE_FILE.mtime)
```

Comment:
This makes me wonder if a `stale_file` helper that takes e.g. `Homebrew::EnvConfig.api_auto_update_secs` and `DATABASE_FILE` as parameters would be a nice Homebrew/brew addition for existing usage and when this gets merged to simplify the code a bit.